src/index.ts

```diff
@@ -0,0 +1,233 @@
+import type { NextFunction, Request, Response } from 'express'
+import type { Gauge, Counter, Histogram } from 'prom-client'
+import type { IncomingMessage, RequestOptions } from 'node:http'
+
+const prometheus = require('prom-client')
+const onFinished = require('on-finished')
+const { env } = require('node:process')
+const http = require('node:http')
+
+const EXPRESS_METRIC_STATUS_CODE = 'status_code'
+const EXPRESS_METRIC_HTTP_METHOD = 'http_method'
+const EXPRESS_METRIC_PATH = 'path'
+
+const customMetrics: CustomMetrics = {}
+const path = new RegExp('^/metrics?$')
+
+let defaultMetricsLabels: DefaultMetricsLabels = {}
+let expressHttpHistogram: Histogram
+let ecsLabelsRequired: boolean = false
+let ecsLabelsValidated: boolean = false
+
+const middleware = (req: Request, res: Response, next: NextFunction) => {
+  if (req.url.match(path)) {
+    if (ecsLabelsValidated || !ecsLabelsRequired) {
+      res.format({
+        'text/plain': async () => {
+          res.set('Content-Type', prometheus.register.contentType)
+          res.end(await prometheus.register.metrics())
+        },
+        'application/json': async () => {
+          res.setHeader('Content-Type', 'application/json')
+          res.end(JSON.stringify(await prometheus.register.getMetricsAsJSON()))
+        },
+        default: async () => {
+          res.set('Content-Type', prometheus.register.contentType)
+          res.end(await prometheus.register.metrics())
+        },
+      })
+    } else {
+      res.setHeader('Content-Type', 'application/json')
+      res.status(501).send(JSON.stringify({ error: 'metrics initialization error' }))
+    }
+    return
+  }
+
+  const expressResponseMetrics: { [key: string]: number | string } = {}
+  const stop = expressHttpHistogram.startTimer(expressResponseMetrics)
+  onFinished(res, () => {
+    expressResponseMetrics[EXPRESS_METRIC_STATUS_CODE] = res.statusCode
+    expressResponseMetrics[EXPRESS_METRIC_HTTP_METHOD] = req.method
+    if (req.route) {
+      expressResponseMetrics[EXPRESS_METRIC_PATH] = req.route.path
+    }
+    stop()
+  })
+  next()
+}
+
+const configure = (opts: MetricsConfigurationOptions = {}) => {
+  if (env.NODE_ENV === 'production') {
+    ecsLabelsRequired = true
+  }
+
+  if (opts.defaultMetricsLabels) {
+    defaultMetricsLabels = opts.defaultMetricsLabels
+    prometheus.register.setDefaultLabels(defaultMetricsLabels)
+  }
+
+  if (opts.fetchECSLabels) {
```

Comment:
Am having a hard time distinguish the need for `ecsLabelsRequired` and `fetchECSLabels`. What's the difference?

Reply:
We could:

- provide a `ECS_CONTAINER_METADATA_URI_V4`, but set `fetchECSLabels` to false
- set `ECS_CONTAINER_METADATA_URI_V4` to null/empty, but set `fetchECSLabels` to true

But what's the point?

Reply:
I thought about this yesterday and figured we could probably settle on enforcing ECS labels if the environment has the URI set rather than relying on the NODE_ENV. Then the need for a configuration option goes away completely.